File: app/src/main/java/jp/juggler/subwaytooter/action/Action_Account.java
package jp.juggler.subwaytooter.action;

import android.app.Dialog;
import android.content.Intent;
import android.net.Uri;
import android.os.Build;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.app.AlertDialog;

import jp.juggler.subwaytooter.ActAccountSetting;
import jp.juggler.subwaytooter.ActMain;
import jp.juggler.subwaytooter.ActPost;
import jp.juggler.subwaytooter.App1;
import jp.juggler.subwaytooter.Column;
import jp.juggler.subwaytooter.Pref;
import jp.juggler.subwaytooter.R;
import jp.juggler.subwaytooter.api.TootApiClient;
import jp.juggler.subwaytooter.api.TootApiResult;
import jp.juggler.subwaytooter.api.TootTask;
import jp.juggler.subwaytooter.api.TootTaskRunner;
import jp.juggler.subwaytooter.dialog.AccountPicker;
import jp.juggler.subwaytooter.dialog.DlgTextInput;
import jp.juggler.subwaytooter.dialog.LoginForm;
import jp.juggler.subwaytooter.table.SavedAccount;
import jp.juggler.subwaytooter.util.LogCategory;
import jp.juggler.subwaytooter.util.Utils;

public class Action_Account {
	private static final LogCategory log = new LogCategory( "Action_Account" );
	
	// アカウントの追加
	public static void add( @NonNull final ActMain activity ){
		
		LoginForm.showLoginForm(
			activity
			, null
			, new LoginForm.LoginFormCallback() {
				
				@Override public void startLogin(
					@NonNull final Dialog dialog
					, @NonNull final String instance
					, final boolean bPseudoAccount
					, final boolean bInputAccessToken
				){
					new TootTaskRunner( activity, true ).run( instance, new TootTask() {
						@Override public TootApiResult background( @NonNull TootApiClient client ){
							if( bPseudoAccount ){
								return client.checkInstance();
							}else{
								String client_name = Pref.pref( activity ).getString( Pref.KEY_CLIENT_NAME, "" );
								return client.authorize1( client_name );
							}
						}
						
						@Override public void handleResult( @Nullable TootApiResult result ){
							if( result == null ) return; // cancelled.
							
							if( result.error != null ){
								String sv = result.error;
								
								// エラーはブラウザ用URLかもしれない
								if( sv.startsWith( "https" ) ){
									
									if( bInputAccessToken ){
										// アクセストークンの手動入力
										DlgTextInput.show( activity, activity.getString( R.string.access_token ), null, new DlgTextInput.Callback() {
											@Override
											public void onOK( Dialog dialog_token, String access_token ){
												activity.checkAccessToken( dialog, dialog_token, instance, access_token, null );
											}
											
											@Override public void onEmptyError(){
												Utils.showToast( activity, true, R.string.token_not_specified );
											}
										} );
									}else{
										// OAuth認証が必要
										Intent data = new Intent();
										data.setData( Uri.parse( sv ) );
										activity.startAccessTokenUpdate( data );
										try{
											dialog.dismiss();
										}catch( Throwable ignored ){
											// IllegalArgumentException がたまに出る
										}
									}
									return;
								}
								
								log.e( result.error );
								
								if( sv.contains( "SSLHandshakeException" )
									&& ( Build.VERSION.RELEASE.startsWith( "7.0" )
									|| ( Build.VERSION.RELEASE.startsWith( "7.1" ) && ! Build.VERSION.RELEASE.startsWith( "7.1." ) ) )
									){
									new AlertDialog.Builder( activity )
										.setMessage( sv + "\n\n" + activity.getString( R.string.ssl_bug_7_0 ) )
										.setNeutralButton( R.string.close, null )
										.show();
									return;
								}
								
								// 他のエラー
								Utils.showToast( activity, true, sv );
							}else{
								
								SavedAccount a = ActionUtils.addPseudoAccount( activity, instance );
								if( a != null ){
									// 疑似アカウントが追加された
									Utils.showToast( activity, false, R.string.server_confirmed );
									int pos = App1.app_state.column_list.size();
									activity.addColumn( pos, a, Column.TYPE_LOCAL );
									
									try{
										dialog.dismiss();
									}catch( Throwable ignored ){
										// IllegalArgumentException がたまに出る
									}
								}
							}
							
						}
					} );
				}
			}
		);
		
	}
	
	// アカウント設定
	public static void setting( @NonNull final ActMain activity ){
		AccountPicker.pick(
			activity
			, true
			, true
			, activity.getString( R.string.account_picker_open_setting )
			, new AccountPicker.AccountPickerCallback() {
				@Override public void onAccountPicked( @NonNull SavedAccount ai ){
					ActAccountSetting.open( activity, ai, ActMain.REQUEST_CODE_ACCOUNT_SETTING );
				}
			} );
	}
	
	// アカウントを選んでタイムラインカラムを追加
	public static void timeline(
		@NonNull final ActMain activity
		, final int pos
		, boolean bAllowPseudo
		, final int type
		, final Object... args
	){
		AccountPicker.pick( activity, bAllowPseudo, true
			, activity.getString( R.string.account_picker_add_timeline_of, Column.getColumnTypeName( activity, type ) )
			, new AccountPicker.AccountPickerCallback() {
				@Override public void onAccountPicked( @NonNull SavedAccount ai ){
					switch( type ){
					default:
						activity.addColumn( pos, ai, type, args );
						break;
					
					case Column.TYPE_PROFILE:
						activity.addColumn( pos, ai, type, ai.id );
						break;
					}
				}
			} );
	}
	
	// 投稿画面を開く。簡易入力があれば初期値はそれになる
	public static void openPost( @NonNull final ActMain activity ){
		openPost( activity, activity.getQuickTootText() );
	}
	
	// 投稿画面を開く。初期テキストを指定する
	public static void openPost(
		@NonNull final ActMain activity
		, @Nullable final String initial_text
	){
		activity.post_helper.closeAcctPopup();
		
		long db_id = activity.getCurrentPostTargetId();
		if( db_id != - 1L ){
			ActPost.open( activity, ActMain.REQUEST_CODE_POST, db_id, initial_text );
		}else{
			AccountPicker.pick(
				activity
				, false
				, true
				, activity.getString( R.string.account_picker_toot )
				, new AccountPicker.AccountPickerCallback() {
					@Override public void onAccountPicked( @NonNull SavedAccount ai ){
						ActPost.open( activity, ActMain.REQUEST_CODE_POST, ai.db_id, initial_text );
					}
				}
			);
		}
	}
}

File: app/src/main/java/jp/juggler/subwaytooter/action/Action_Account.kt
package jp.juggler.subwaytooter.action

import android.os.Build
import jp.juggler.subwaytooter.*
import jp.juggler.subwaytooter.actmain.addColumn
import jp.juggler.subwaytooter.actmain.afterAccountVerify
import jp.juggler.subwaytooter.actmain.defaultInsertPosition
import jp.juggler.subwaytooter.api.*
import jp.juggler.subwaytooter.api.auth.Auth2Result
import jp.juggler.subwaytooter.api.auth.AuthBase
import jp.juggler.subwaytooter.api.entity.*
import jp.juggler.subwaytooter.column.ColumnType
import jp.juggler.subwaytooter.dialog.*
import jp.juggler.subwaytooter.dialog.DlgCreateAccount.Companion.showUserCreateDialog
import jp.juggler.subwaytooter.dialog.LoginForm.Companion.showLoginForm
import jp.juggler.subwaytooter.table.SavedAccount
import jp.juggler.subwaytooter.util.LinkHelper
import jp.juggler.subwaytooter.util.openBrowser
import jp.juggler.util.coroutine.launchMain
import jp.juggler.util.data.JsonObject
import jp.juggler.util.data.encodePercent
import jp.juggler.util.log.LogCategory
import jp.juggler.util.log.showToast
import jp.juggler.util.network.toFormRequestBody
import jp.juggler.util.network.toPost
import jp.juggler.util.ui.dismissSafe

private val log = LogCategory("Action_Account")

// Androidでは \w や \d がUnicode文字にマッチしてしまうので、IDEの警告を無視する
@Suppress("RegExpSimplifiable")
private val mailRegex =
    """\A[a-z0-9_+&*-]+(?:\.[a-z0-9_+&*-]+)*@(?:[a-z0-9-]+\.)+[a-z]{2,12}\z""".toRegex(
        RegexOption.IGNORE_CASE
    )

fun isAndroid7TlsBug(errorText: String) =
    if (!errorText.contains("SSLHandshakeException")) {
        false
    } else {
        val release = Build.VERSION.RELEASE
        when {
            release.startsWith("7.1.") -> false // 含まない 7.1.x
            release.startsWith("7.1") -> true   // 含む 7.1
            release.startsWith("7.0") -> true   // 含む 7.0
            else -> false
        }
    }

/**
 * サイドメニューで「アカウントの追加」を選ぶと呼び出される。
 * - サーバ名とアクションを指定するダイアログを開く。
 * - 選択されたアクションに応じて分岐する。
 */
fun ActMain.accountAdd() {
    @Suppress("UNUSED_ANONYMOUS_PARAMETER")
    showLoginForm { dialogHost, apiHost, serverInfo, action ->
        launchMain {
            try {
                when (action) {
                    LoginForm.Action.Login -> {
                        val authUri = runApiTask2(apiHost) { it.authStep1() }
                        log.i("authUri=$authUri")
                        openBrowser(authUri)
                        dialogHost.dismissSafe()
                    }

                    LoginForm.Action.Pseudo -> {
                        val tootInstance = runApiTask2(apiHost) { TootInstance.getOrThrow(it) }
                        addPseudoAccount(apiHost, tootInstance)?.let { a ->
                            showToast(false, R.string.server_confirmed)
                            addColumn(defaultInsertPosition, a, ColumnType.LOCAL, protect = true)
                            dialogHost.dismissSafe()
                        }
                    }
//                    LoginForm.Action.Create ->
//                        createUser(apiHost, serverInfo) { dialogHost.dismissSafe() }
                    LoginForm.Action.Token ->
                        accessTokenPrompt(apiHost) { dialogHost.dismissSafe() }
                }
            } catch (ex: Throwable) {
                showApiError(ex)
            }
        }
    }
}

private suspend fun ActMain.createUser(
    apiHost: Host,
    serverInfo: TootInstance?,
    onComplete: (() -> Unit)? = null,
) {
    serverInfo ?: error(
        getString(
            R.string.user_creation_not_supported,
            apiHost.pretty,
            "(unknown)",
        )
    )

    fun TootApiClient.authUserCreate() =
        AuthBase.findAuthForCreateUser(this, serverInfo)
            ?: error(
                getString(
                    R.string.user_creation_not_supported,
                    apiHost.pretty,
                    serverInfo.instanceType.toString(),
                )
            )

    // クライアント情報を取得。サーバ種別によってはユーザ作成ができないのでエラーとなる
    val clientInfo = runApiTask2(apiHost) {
        it.authUserCreate().prepareClient(serverInfo)
    }

    showUserCreateDialog(apiHost) { dialogCreate, params ->
        launchMain {
            try {
                val auth2Result = runApiTask2(apiHost) {
                    it.authUserCreate().createUser(clientInfo = clientInfo, params = params)
                }
                if (afterAccountVerify(auth2Result)) {
                    dialogCreate.dismissSafe()
                    onComplete?.invoke()
                }
            } catch (ex: Throwable) {
                showApiError(ex)
            }
        }
    }
}

/**
 * アクセストークンを手動入力する。
 *
 * @param apiHost アクセストークンと関係のあるサーバのAPIホスト
 * @param onComplete 非nullならアカウント認証が終わったタイミングで呼ばれる
 */
// アクセストークンの手動入力(更新)
suspend fun ActMain.accessTokenPrompt(
    apiHost: Host,
    onComplete: (() -> Unit)? = null,
) {
    showTextInputDialog(
        title = getString(R.string.access_token_or_api_token),
        initialText = null,
        onEmptyText = { showToast(true, R.string.token_not_specified) },
    ) { text ->
        try {
            val accessToken = text.trim()
            val auth2Result = runApiTask2(apiHost) { client ->
                val ti = TootInstance.getExOrThrow(client, forceAccessToken = accessToken)

                val tokenJson = JsonObject()

                val userJson = client.verifyAccount(
                    accessToken,
                    outTokenInfo = tokenJson, // 更新される
                    misskeyVersion = ti.misskeyVersionMajor
                )
                val parser = TootParser(this, linkHelper = LinkHelper.create(ti))

                Auth2Result(
                    tootInstance = ti,
                    tokenJson = tokenJson,
                    accountJson = userJson,
                    tootAccount = parser.account(userJson)
                        ?: error("can't parse user information."),
                )
            }
            when (afterAccountVerify(auth2Result)) {
                false -> false
                else -> {
                    onComplete?.invoke()
                    true
                }
            }
        } catch (ex: Throwable) {
            showApiError(ex)
            false
        }
    }
}

// アカウント設定
fun ActMain.accountOpenSetting() {
    launchMain {
        pickAccount(
            bAllowPseudo = true,
            bAuto = true,
            message = getString(R.string.account_picker_open_setting)
        )?.let {
            arAccountSetting.launch(ActAccountSetting.createIntent(this@accountOpenSetting, it))
        }
    }
}

fun ActMain.accountResendConfirmMail(accessInfo: SavedAccount) {
    DlgConfirmMail(
        this,
        accessInfo
    ) { email ->
        launchMain {
            runApiTask(accessInfo) { client ->
                email?.let {
                    if (!mailRegex.matches(it)) {
                        return@runApiTask TootApiResult("email address is not valid.")
                    }
                }

                client.request(
                    "/api/v1/emails/confirmations",
                    ArrayList<String>().apply {
                        if (email != null) add("email=${email.encodePercent()}")
                    }.joinToString("&").toFormRequestBody().toPost()
                )
            }?.let { result ->
                when (val error = result.error) {
                    null -> showToast(true, R.string.resend_confirm_mail_requested)
                    else -> showToast(true, error)
                }
            }
        }
    }.show()
}
